src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -405,4 +411,28 @@ public void EnumerateLinks(Activity activity, int maxLinks)
             }
         }
     }
+
+    private struct ActivitySourceTagsEnumerationState
+    {
+        public SdkLimitOptions SdkLimitOptions;
+
+        public InstrumentationScope InstrumentationScope;
+
+        public void EnumeratTags(ActivitySource activitySource, int maxTags)
+        {
+            var scopeAttributes = this.InstrumentationScope.Attributes;
+
+            foreach (var tag in activitySource.Tags ?? [])
```

Comment:
ActivitySource.Tags is IEnumerable so this `foreach` is going to allocate an enumerator. It seems to be a IReadOnlyList behind the scenes: https://github.com/dotnet/runtime/blob/cb256e20e4e4db764f7693d4ed5de932a67ec6e9/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L48

You _could_ try to optimize for that if you want to avoid the allocation:

```csharp
if (activitySource.Tags is IReadOnlyList<KeyValuePair<string, object?>> list)
{
   for (int i = 0; i < list.Count; i++)
   {
        var attribute = list[i];
        AddAttribute(..., attribute);
   }
}
else
{
   foreach (var attribute in activitySource.Tags ?? [])
   {
        AddAttribute(..., attribute);
   }
}
```

Reply:
There are 5 parameters to pass to `AddAttribute(..., attribute)`, felt better to repeat the code as it is 2 lines.